### PR TITLE
Improve CPU count detection in cgroup environments and fix CI

### LINKED ISSUE #5809: 🔍 Improve CPU count detection in cgroup environments
### Current problem

PyLint [currently uses](https://github.com/PyCQA/pylint/blob/8c0062f5ac0cfd80a74568e36d1e68e5a128c7f5/pylint/lint/run.py#L22-L30) `sched_getaffinity` or `multiprocessing.cpu_count` to identify the number of CPUs available to the system, falling back to `1` if neither function is available. This doesn't work properly in cgroup environments, where CPU limits may be imposed as a fraction of total CPU time available to the set of processes.

### Desired solution

It would be great if PyLint were to automatically detect the fraction of CPUs available to it when determining how many subprocesses to launch. Currently in our Kubernetes-based CI system, PyLint is launching 16 processes which all compete for a single CPU's worth of time. This ultimately slows down the checking process, which we want to be fast 🙂 

Alternatively, it would be great to be able to set an environment variable to control the number of subprocesses to launch. This would make it possible for us set the number of subprocesses in the same place where we set the number of CPUs available to the CI process.

As a workaround, we're going to modify our launching script to compute the fraction of CPUs available to the process and pass that using the `-j` argument to PyLint.

### Additional context

Determining the amount of CPU available to a Python program is a somewhat widespread problem, here are some related resources:
* Python issue requesting this ability in the standard library: https://bugs.python.org/issue36054
* Message showing how to manually query cgroup quotas: https://bugs.python.org/msg353690
* Conan PR using similar logic to load cgroup quota information: https://github.com/conan-io/conan/pull/5466/files
* Utility using `LD_PRELOAD` to hijack CPU count requests to return computed CPU availability: https://github.com/agile6v/container_cpu_detection

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

After a rough Friday afternoon I think I have finally found the issue: our CI runs in a docker container.
#5809 already raised an issue about this and our approach at determining the number of available CPUs. I have adopted a suggestion to fix this issue from the Python bug report found at https://bugs.python.org/issue36054.

I have tested this and `_query_cpu` returns `1` for us in the CI, both on Linux <s>and on Windows</s>. As you will see in this CI run the test marked with the new `needs_two_cores` mark are indeed skipped.

This would thus close #5809 as well as fix our CI issues.

<s>One issue that results from this is that these tests won't be run on our CI. They will still run locally (I checked this on my own local Mac), but won't be run in the CI. I don't immediately know a solution for this. The Github Actions runners simply don't really allow us to test our approach at parallelisation I think...</s>

Edit: I misread and on Windows the jobs do keep running. Perhaps that was just a fluke and it changes based on the runner? That would explain some of the variance we have been seeing. Anyway, seems as if the test will run at least some of the time. So that is probably good enough for now?

### CODE REVIEW

**Comment by DanielNoord** [comment]
I'm trying to turn on `3.11` as there is finally a release available in https://github.com/DanielNoord/pylint/runs/6037333512?check_suite_focus=true but I think we're getting stuck here even more often.

Probably can't start testing on `3.11` until this lands.

**Review by Pierre-Sassoulas — COMMENTED** [review]
I'm not qualified to review query_cpu and I don't think I'll be able to be in a reasonable time, but I guess you did your research properly and that detecting CPU's inside a docker is can't be done without some hacks 😄 The principle of the mark when two cores are required is consensual so 👍 for that. I guess the worst that can happen if qury_cpu do not work is that it use the wrong number of CPU when inside a docker ?

**Comment by DanielNoord** [comment]
> I'm not qualified to review query_cpu and I don't think I'll be able to be in a reasonable time, but I guess you did your research properly and that detecting CPU's inside a docker is can't be done without some hacks 😄 

Do we have somebody who would be better suited to do this?

> The principle of the mark when two cores are required is consensual so 👍 for that. I guess the worst that can happen if qury_cpu do not work is that it use the wrong number of CPU when inside a docker ?

Yeah! Or any other environment that uses these files.

**Review by Pierre-Sassoulas — APPROVED** [review]

File: pylint/lint/run.py [hunk 1]
```diff
@@ -36,15 +37,51 @@
     multiprocessing = None  # type: ignore[assignment]
 
 
+def _query_cpu() -> int | None:
+    """Try to determine number of CPUs alloted in a docker container.
```
Comment (jacobtylerwalls) [inline pylint/lint/run.py:41]:
```suggestion
    """Try to determine number of CPUs allotted in a docker container.
```

File: pylint/lint/run.py [hunk 2]
```diff
@@ -36,15 +37,51 @@
...
+
+    This is based on discussion and coped from suggestions in
```
Comment (jacobtylerwalls) [inline pylint/lint/run.py:43]:
```suggestion
    This is based on discussion and copied from suggestions in
```

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by jacobtylerwalls** [comment]
Thanks! If this ever gets brittle/stops working because the file locations change, we could look into the 3rd party packages on the BPO issue.

**Pull request merged**